Add multi select pattern documentation (#1938)
hashicorp/design-system: website/docs/patterns/table-multi-select/partials/interaction-concepts/interaction-concepts.md

When combining a Multi-select table with filtering, sorting, or pagination, you may encounter complex scenarios regarding how selected results should be persisted across pages and what the user expects when traversing large data sets.

These interaction guidelines are not fully comprehensive. They should instead be considered as conversation starters between design and engineering on how best to optimize the user experience.

!!! Info

If you run into unique scenarios that these guidelines don’t account for, we would love to hear about them through our normal [support](/about/support) channels.
!!!

## Consistent communication

> The most important method to ensure users are interacting with the results they intend to is to clearly communicate what they have selected, or even that a selection of results exists, **at all times**.

Reviewer: ```suggestion
To ensure users are interacting with the results they intend to, it is vital to consistently communicate their selections, or the existence of a selection, **at all times**.
```

> This can help to prevent unnecessary deletions of results or making unintended changes if what is currently selected is unclear.

Reviewer: ```suggestion
This can help to prevent unintended deletion of data or unintended changes.
```

Reviewer: These two sentences could probably still use a bit of massaging to further clarify but I attempted to simplify them a bit...

Author: I've done a bit to combine these sentences and add a bit of clarifying imagery and information.

A user is viewing a data set consisting of 10,000 results that is presented in a Table. In order to make changes to specific results the user may take the following actions:

1. **Filter** the Table to display only a subset of the results.
2. **Sort** the Table by a specific parameter to order the results in a meaningful way.
3. Traverse through multiple pages of the Table through **Pagination**.

If not handled in a logical manner, how the selection is persisted across all of these actions can be confusing and open the end-user up to making unintended changes if what they have selected is persisted (or not persisted).

### Within a paginated Table

If a user makes a selection on one page of a paginated Table and progresses to the next page, persisting this selection can be communicated by the [selected count](/patterns/multi-select-patterns?tab=guidelines#selected-count) and can be helpful in scenarios where much of the data is similar or overlaps.

<video loop controls width="100%">
    <source
        src="/assets/patterns/table-multi-select/multi-select-pagination-interaction.mp4"
        type="video/mp4"
    />
</video>

### When filtering Table results

> Filtering a data set implies removing or hiding results from rendering within a Table; a **reductive** action. In this scenario, performing a reductive action will impact multiple areas of the interface and surface a number of considerations for persisting the selection.

Reviewer: This sentence seems repetitive and is difficult for me to understand. Could you simplify it a bit to clarify?

Author: I've done a bit of massaging here to clean this up.

!!! Info

> Consider these topics conversation starters to help make an informed decision around the user experience, the answers may vary depending on product architecture and broader application strategy.

Reviewer: ```suggestion
Consider these topics as conversation starters to help make an informed decision about the user experience. The answers may vary depending on product architecture and broader application strategy.
```

!!!

- In a Multi-select pattern, should the [total count](/patterns/multi-select-patterns?tab=guidlines#total-count) change to reflect a reduction in the number of visible results?
- In Pagination, is the number of results reduced to reflect the those excluded by the filter?
- If a selection has been made prior to filtering, is the selected state persisted even if the filtered results are excluded from the data set?
- Are there any parameters in the data set that are variable in nature? E.g., status that is prone to changing if a cluster is starting up or spinning down.

> In all of these scenarios it is most important to **clearly and consistently communicate** what results have been selected and what impact a change will have on them.

Reviewer: ```suggestion
In all of these scenarios it is important to **clearly and consistently communicate** what results have been selected and what impact a change will have on them.
```

When combining a Multi-select table with filtering, sorting, or pagination, you may encounter complex scenarios regarding how selected results should be persisted across pages and what the user expects when traversing large data sets.

These interaction guidelines are not fully comprehensive. They should instead be considered as conversation starters between design and engineering on how best to optimize the user experience.

!!! Info

If you run into unique scenarios that these guidelines don’t account for, we would love to hear about them through our normal [support](/about/support) channels.
!!!

## Consistent communication

To help prevent unnecessary editing or deleting and to ensure users are interacting with the results they intend to, it is vital to consistently communicate their selections, or the existence of a selection, **at all times**.

> Persisting this communication in the selected count can help to avoid mishaps, but sometimes it might be necessary to confirm with the user prior to making changes that cannot be reversed.

Reviewer: ```suggestion
Persisting this communication in the selected count can help to avoid mishaps, but it may also be necessary to confirm with the user prior to making changes that cannot be reversed.
```

> ![Example of a double confirmation](/assets/patterns/table-multi-select/multi-select-double-confirmation.png)

Author: Yes, that's a better title I think, I think I was intending to differentiate from the other example, but "Are you sure?" isn't a great title for a Modal.

## Selection scope

> _Where_ a selection is made in the UI can imply differences in how the selection is scoped, and at a very basic level can broken down into three different hierarchies.

Reviewer: ```suggestion
_Where_ a selection is made in the UI can imply differences in how the selection is scoped, and can be broken down into three different levels of hierarchy.
```

> 1. **Global scope:** selection applies to all rows in the table and is made via the bulk selection in the Multi-select pattern.

Reviewer: ```suggestion
1. **Global scope:** Selection applies to all rows in the table and is made via bulk selection in the Multi-select pattern.
```

> 2. **Page-level scope:** selection applies to all rows in the current page of a paginated table and is made via the Checkbox in the Table header.

Reviewer: ```suggestion
2. **Page-level scope:** Selection applies to all rows in the current page of a paginated table and is made via the Checkbox in the Table header.
```

> 3. **Row-level scope:** selection applies only to a single row, result, or item in a Table, and is made via the checkbox at the start of the row.

Reviewer: ```suggestion
3. **Row-level scope:** Selection applies only to a single row, result, or item in a Table and is made via the checkbox at the start of the row.
```

> ![Selection scope](/assets/patterns/table-multi-select/selection-scope.png)

Reviewer 2: In this image/representation the table doesn't have a pagination, so it's confusing what "page-level" scope refers to

### Within a paginated Table

If a user makes a selection on one page of a paginated Table and progresses to the next page, persisting this selection can be communicated by the [selected count](/patterns/multi-select-patterns?tab=guidelines#selected-count) and can be helpful in scenarios where much of the data is similar or overlaps.

<video loop controls width="100%">
    <source
        src="/assets/patterns/table-multi-select/multi-select-pagination-interaction.mp4"
        type="video/mp4"
    />
</video>

### When filtering Table results

Filtering a data set implies removing or hiding results from rendering within a Table (a **reductive** action) and will impact multiple areas of the interface while raising number of considerations around persisting the selection.

!!! Info

Consider these topics as conversation starters to help make an informed decision about the user experience. The answers may vary depending on product architecture and broader application strategy.
!!!

- In a Multi-select pattern, should the [total count](/patterns/multi-select-patterns?tab=guidlines#total-count) change to reflect a reduction in the number of visible results?

> - In Pagination, is the number of results reduced to reflect the those excluded by the filter?

Reviewer: ```suggestion
- In Pagination, is the number of results reduced to reflect those excluded by the filter?
```
